src/mito2/src/cache/index.rs

```diff
@@ -59,43 +60,58 @@ where
         offset: u64,
         size: u32,
     ) -> index::inverted_index::error::Result<Vec<u8>> {
-        let range = offset as usize..(offset + size as u64) as usize;
-        if let Some(cached) = self.cache.get_index(IndexKey {
-            file_id: self.file_id,
-        }) {
-            CACHE_HIT.with_label_values(&[INDEX_CONTENT_TYPE]).inc();
-            Ok(cached[range].to_vec())
-        } else {
-            let mut all_data = Vec::with_capacity(1024 * 1024);
-            self.inner.read_all(&mut all_data).await?;
-            let result = all_data[range].to_vec();
-            self.cache.put_index(
-                IndexKey {
-                    file_id: self.file_id,
-                },
-                Arc::new(all_data),
-            );
-            CACHE_MISS.with_label_values(&[INDEX_CONTENT_TYPE]).inc();
-            Ok(result)
+        let indexes = IndexKey::index(self.file_id, offset, size, self.cache.page_size);
+        // Size is 0, return empty data.
+        if indexes.is_empty() {
+            return Ok(Vec::new());
+        }
+        let mut data = Vec::with_capacity(size as usize);
+        // Safety: indexes is not empty.
+        let first_page_id = indexes[0].page_id;
+        let last_page_id = indexes[indexes.len() - 1].page_id;
+        for index in indexes {
+            let page = match self.cache.get_index(&index) {
+                Some(page) => {
+                    CACHE_HIT.with_label_values(&[INDEX_CONTENT_TYPE]).inc();
+                    page
+                }
+                None => {
+                    let page = self
+                        .inner
+                        .range_read(
+                            index.page_id * (self.cache.page_size as u64),
+                            self.cache.page_size as u32,
+                        )
+                        .await?;
+                    self.cache.put_index(index.clone(), Arc::new(page.clone()));
+                    CACHE_MISS.with_label_values(&[INDEX_CONTENT_TYPE]).inc();
+                    Arc::new(page)
```

Comment:
We can read all cache-hit pages first, and fetch all cache-missed pages via `RangeReader::read_vec`.